File: go/0146_lru_cache.go
func Constructor(capacity int) LRUCache {
	m := make(map[int]*CacheNode)
	c := LRUCache{
		Cap: capacity,
		Map: m,
	}
	return c
}

func (this *LRUCache) Get(key int) int {
	found, ok := this.Map[key]
	if !ok {
		return -1
	}
	if this.Head == found {
		return found.Val
	}
	if this.Tail == found {
		this.Tail = found.Prev
	}
	// move found to head
	if found.Next != nil {
		found.Next.Prev = found.Prev
	}
	if found.Prev != nil {
		found.Prev.Next = found.Next
	}
	this.Head.Prev, found.Nex = found, this.Head
	this.Head = found
	return found.Val
}

func (this *LRUCache) Put(key int, value int) {
	found, ok := this.Map[key]
	if ok {
		found.Val = value
		_ = this.Get(found.Key) // to move found node to head
		return
	}

	// add to head
	n := &CacheNode{Key: key, Val: value}

	if len(this.Map) == 0 {
		this.Tail = n
	} else {
		this.Head.Prev, n.Next = n, this.Head
	}
	this.Map[key], this.Head = n, n
	if this.Cap == this.Len {
		delete(this.Map, this.Tail.Key)
		this.Len, this.Tail = this.Len+1, this.Tail.Prev
	}
	this.Len++
}

type CacheNode struct {
	Next *CacheNode
	Prev *CacheNode
	Key  int
	Val  int
}

type LRUCache struct {
	Cap  int
	Len  int
	Head *CacheNode
	Tail *CacheNode
	Map  map[int]*CacheNode
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * obj := Constructor(capacity);
 * param_1 := obj.Get(key);
 * obj.Put(key,value);
 */

File: python/0146_lru_cache.py
class LRUCache:

    def __init__(self, capacity: int):
        self.head = None
        self.tail = None
        self.map = {}
        self.cur = 0
        self.cap = capacity

    def get(self, key: int) -> int:
        res = self._get(key)
        if res is None:
            return -1
        else:
            return res.val

    def put(self, key: int, value: int) -> None:
        res = self._get(key)
        if res is not None:
            res.val = value
            return

        node = LRUNode(key, value)
        self.map[key] = node

        if self.head is None:
            self.head = node
            self.tail = node
            self.cur = 1
            return

        self.head.prev = node
        node.next = self.head
        self.head = node

        if self.cur < self.cap:
            self.cur += 1
        else:
            tail = self.tail
            prev = tail.prev
            prev.next = None
            self.tail = prev
            del self.map[tail.key]

    def _get(self, key):
        if key not in self.map:
            return None

        node = self.map[key]
        if self.head == node:
            return node
        if self.tail == node:
            prev = self.tail.prev
            self.tail = prev
            prev.next = None
            node.prev = None

            node.next = self.head
            self.head.prev = node
            self.head = node
            return node

        prev = node.prev
        next = node.next
        prev.next = next
        next.prev = prev

        self.head.prev = node
        node.next = self.head
        self.head = node
        return node


class LRUNode:
    def __init__(self, key, val):
        self.key = key
        self.val = val
        self.next = None
        self.prev = None

# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)